components/layout_2020/flexbox/layout.rs

```diff
@@ -2,32 +2,1090 @@
+    }
+
+    fn sides_to_flow_relative<T>(&self, x: FlexRelativeSides<T>) -> Sides<T> {
+        self.main_start_cross_start_sides_are
+            .sides_to_flow_relative(x)
+    }
+
+    fn rect_to_flow_relative(
+        &self,
+        base_rect_size: FlexRelativeVec2<Length>,
+        rect: FlexRelativeRect<Length>,
+    ) -> Rect<Length> {
+        super::geom::rect_to_flow_relative(
+            self.flex_axis,
+            self.main_start_cross_start_sides_are,
+            base_rect_size,
+            rect,
+        )
+    }
+}
 
 impl FlexContainer {
+    pub fn inline_content_sizes(&self) -> ContentSizes {
+        // FIXME: implement this. The spec for it is the same as for "normal" layout:
+        // https://drafts.csswg.org/css-flexbox/#layout-algorithm
+        // … except that the parts that say “the flex container is being sized
+        // under a min or max-content constraint” apply.
+        ContentSizes::zero() // Return an incorrect result rather than panic
+    }
+
+    /// https://drafts.csswg.org/css-flexbox/#layout-algorithm
     pub(crate) fn layout(
         &self,
         layout_context: &LayoutContext,
         positioning_context: &mut PositioningContext,
         containing_block: &ContainingBlock,
         tree_rank: usize,
     ) -> IndependentLayout {
-        // FIXME
-        let _ = layout_context;
-        let _ = positioning_context;
-        let _ = containing_block;
-        let _ = tree_rank;
+        // Actual length may be less, but we guess that usually not by a lot
+        let mut flex_items = Vec::with_capacity(self.children.len());
+
+        // Absolutely-positioned children of the flex container may be interleaved
+        // with flex items. We need to preserve their relative order for correct painting order,
+        // which is the order of `Fragment`s in this function’s return value.
+        let original_order_with_absolutely_positioned = self
+            .children
+            .iter()
+            .enumerate()
+            .map(|(tree_rank, arcrefcell)| {
+                let borrowed = arcrefcell.borrow_mut();
+                match &*borrowed {
+                    FlexLevelBox::OutOfFlowAbsolutelyPositionedBox(absolutely_positioned) => {
+                        Ok(absolutely_positioned.clone())
+                    },
+                    FlexLevelBox::FlexItem(_) => {
+                        let item = AtomicRefMut::map(borrowed, |child| match child {
+                            FlexLevelBox::FlexItem(item) => item,
+                            _ => unreachable!(),
+                        });
+                        flex_items.push((tree_rank, item));
+                        Err(())
+                    },
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let mut content_block_size_option_dance = None;
+        let fragments =
+            positioning_context.adjust_static_positions(tree_rank, |positioning_context| {
```

Comment:
Why does this need to be done? What does it do?

Reply:
The `BoxFragment::content_rect::start_corner` position of a fragment is a point in a coordinate space whose origin is the start corner of the *parent* in the in the fragment tree. Using relative coordinates like this allows translating an entire sub-tree by modifying the coordinates of its root fragment.

When hoisting an absolutely-positioned box "up" the tree, we’re effectively moving to another coordinate space. `HoistedAbsolutelyPositionedBox::box_offsets` may contain coordinates that will contribute to the position of the eventual fragment. `adjust_static_positions` effectively converts them to another coordinate space.